### PR TITLE
[SWPWA-431] Appropriately inform the user when they're offline - offline notice

### PR DESCRIPTION
**Pull request by dmitrijs-voronovs**
* new `OfflineNotice` component
* new `Offline` Store
* cmsPage, PDP, PLP support for **BIG** offline notice
* Hedaer support for **small** offline notice
* in styles add padding for offline notice

Desktop:
Selection_212
Selection_214

Selection_211

Mobile:
Selection_213

Selection_208

### CODE REVIEW

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * See LICENSE for license details.
+ *
+ * @license OSL-3.0 (Open Software License ("OSL") v. 3.0)
+ * @package scandipwa/base-theme
+ * @link https://github.com/scandipwa/base-theme
+ */
+
+:root {
+    --diameter: 1;
+    --offline-notice-height: 0px;
```
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

File: src/app/component/MenuOverlay/MenuOverlay.style.scss [hunk 2]
```diff
@@ -129,7 +129,7 @@
             left: 0;
             width: 100%;
             height: 100%;
-            padding-top: var(--header-height);
+            padding-top: calc(var(--header-height) + var(--offline-notice-height));;
```
Comment (Danpiel) [inline src/app/component/MenuOverlay/MenuOverlay.style.scss]: Codacy Issue found: [Unexpected extra semicolon (no-extra-semicolons)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss (hunk 1 above)
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss (hunk 1 above)
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

File: src/app/component/OfflineNotice/OfflineNotice.container.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * @license OSL-3.0 (Open Software License ("OSL") v. 3.0)
+ * @package scandipwa/base-theme
+ * @link https://github.com/scandipwa/base-theme
+ */
+
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+import { connect } from 'react-redux';
+import { showOfflineNotice, setBigOfflineNotice } from 'Store/Offline';
+import { withRouter } from 'react-router';
```
Comment (alfredsgenkins) [inline src/app/component/OfflineNotice/OfflineNotice.container.js]: Order imports

File: src/app/component/OfflineNotice/OfflineNotice.container.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+        if (!isPage) {
+            window.removeEventListener('online', this.handleNetworkChange);
+            window.removeEventListener('offline', this.handleNetworkChange);
+        }
+    }
+
+    handleNetworkChange = () => {
+        const {
+            isBig, showOfflineNotice, setBigOfflineNotice
```
Comment (alfredsgenkins) [inline src/app/component/OfflineNotice/OfflineNotice.container.js]: FYI perefer on new line each :)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    --offline-notice-display: none;
+}
+
+.OfflineNotice {
+    --fill-color: var(--primary-info-color);
+    --line-color: black;
+    display: var(--offline-notice-display);
```
Comment (alfredsgenkins) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Separate css custom variables with enter.

File: src/app/route/CategoryPage/CategoryPage.container.js [hunk 6]
```diff
@@ -115,23 +121,32 @@ export class CategoryPageContainer extends PureComponent {
...
         } else {
             this._onCategoryUpdate();
         }
     }
 
     componentDidUpdate(prevProps) {
-        const { category: { id } } = this.props;
+        const {
+            category: { id }, isOffline
+        } = this.props;
```
Comment (alfredsgenkins) [inline src/app/route/CategoryPage/CategoryPage.container.js]: Please separate with enter from next const declaration.

File: src/app/route/CategoryPage/CategoryPage.container.js [hunk 7]
```diff
@@ -147,6 +162,13 @@ export class CategoryPageContainer extends PureComponent {
         setQueryParams({ sortDirection }, location, history);
     }
 
+    setOfflineNoticeSize = () => {
+        const { setBigOfflineNotice, isInfoLoading } = this.props;
+
+        if (isInfoLoading) setBigOfflineNotice(true);
```
Comment (alfredsgenkins) [inline src/app/route/CategoryPage/CategoryPage.container.js]: Prefer `{ ... }`

File: src/app/route/CmsPage/CmsPage.container.js [hunk 8]
```diff
@@ -80,6 +97,33 @@ export class CmsPageContainer extends DataContainer {
...
+            this.requestPage(urlKey || urlParam);
+        }
+    }
+
+    setOfflineNoticeSize = () => {
+        const { setBigOfflineNotice } = this.props;
+        const { isLoading } = this.state;
+
+        if (isLoading) setBigOfflineNotice(true);
+        else setBigOfflineNotice(false);
```
Comment (alfredsgenkins) [inline src/app/route/CmsPage/CmsPage.container.js]: Prefer `{ ... }`

File: src/app/route/ProductPage/ProductPage.container.js [hunk 9]
```diff
@@ -186,13 +191,15 @@ export class ProductPageContainer extends PureComponent {
...
+        const { isOffline, setBigOfflineNotice } = this.props;
         const dataSource = this._getDataSource();
 
         if (Object.keys(dataSource).length) {
             this._updateBreadcrumbs(dataSource);
             this._updateHeaderState(dataSource);
             this._updateNavigationState();
-        }
+            if (isOffline) setBigOfflineNotice(false);
+        } else if (isOffline) setBigOfflineNotice(true);
```
Comment (alfredsgenkins) [inline src/app/route/ProductPage/ProductPage.container.js]: Prefer `{ ... }`

File: src/app/route/index.js [hunk 10]
```diff
@@ -220,6 +232,24 @@ class AppRouter extends PureComponent {
         );
     }
 
+    renderMainItems() {
+        const { isOffline, isBig } = this.props;
+
+        if (isOffline && isBig) {
+            return <OfflineNotice isPage />;
```
Comment (alfredsgenkins) [inline src/app/route/index.js]: Maybe let's render it the same place as the error notice? Is it possible?

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+ * See LICENSE for license details.
+ *
+ * @license OSL-3.0 (Open Software License ("OSL") v. 3.0)
+ * @package scandipwa/base-theme
+ * @link https://github.com/scandipwa/base-theme
+ */
+
+:root {
+    --diameter: 1;
+    --offline-notice-height: 0px;
```
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

File: src/app/route/index.js (hunk 10 above)
Comment (dmitrijs-voronovs) [reply src/app/route/index.js]: It is possible, but we will be required to copy rendering of header and footer. Should I move it then?

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss (hunk 11 above)
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

**Comment by alfredsgenkins** [comment]
Why so many conflicts? Please take a look :)

**Comment by Danpiel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443&bid=16817423)

File: src/app/component/OfflineNotice/OfflineNotice.style.scss (hunk 11 above)
Comment (Danpiel) [inline src/app/component/OfflineNotice/OfflineNotice.style.scss]: Codacy Issue found: [Unexpected unit (length-zero-no-unit)](https://app.codacy.com/manual/ScandiPWA/base-theme/pullRequest?prid=5115443)

**Comment by dmitrijs-voronovs** [comment]
Ready!

**Pull request merged**